src/components/common/dialog.tsx

```diff
@@ -0,0 +1,1550 @@
+    /* eslint-disable no-use-before-define */
+    #dialog: DialogState<unknown>;
+    /* eslint-enable */
+    #getter: () => T;
+    #setter: (val: T, is_aux: boolean) => void;
+    #path: string;
+
+    constructor(
+        dialog: DialogState<unknown>,
+        getter: () => T,
+        setter: (val: T, is_aux: boolean) => void,
+        path: string
+    ) {
+        this.#dialog = dialog;
+        this.#getter = getter;
+        this.#setter = setter;
+        this.#path = path;
+    }
+
+    validation_text(): string | undefined {
+        return this.#dialog._get_validation(this.#path);
+    }
+
+    get(): T {
+        return this.#getter();
+    }
+
+    set(val: T): void {
+        this.#setter(val, false);
+    }
+
+    set_aux(val: T): void {
+        this.#setter(val, true);
+    }
+
+    id(tag: string = "field"): string {
+        return "dialog-" + tag + "-" + this.#path;
+    }
+
+    map<X>(func: (val: DialogValue<ArrayElement<T>>, index: number) => X): X[] {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            return val.map((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
+        } else
+            return [];
+    }
+
+    forEach(func: (val: DialogValue<ArrayElement<T>>, index: number) => void): void {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            val.forEach((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
+        }
+    }
+
+    remove(index: number) {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            for (let j = index; j < val.length - 1; j++)
+                this.#dialog._rename_validation_state(this.#path, j + 1, j);
+            this.set(toSpliced(val, index, 1) as T);
+        }
+    }
+
+    add(item: ArrayElement<T>) {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            this.set(val.concat(item) as T);
+        }
+    }
+
+    sub<K extends keyof T>(tag: K, update_func?: ((val: T[K]) => void) | undefined): DialogValue<T[K]> {
+        return new DialogValue<T[K]>(
+            this.#dialog,
+            () => this.get()[tag],
+            (val, is_aux) => {
+                const container = this.get();
+                if (Array.isArray(container) && typeof tag == "number")
+                    this.#setter(toSpliced(container, tag, 1, val) as T, is_aux);
+                else
+                    this.#setter({ ...container, [tag]: val }, is_aux);
+                if (update_func)
+                    update_func(val);
+            },
+            this.#path ? this.#path + "." + String(tag) : String(tag)
+        );
+    }
+
+    at<TT extends T>(witness: TT): DialogValue<TT> {
+        cockpit.assert(Object.is(witness, this.get()));
+        return new DialogValue<TT>(
+            this.#dialog,
+            () => this.get() as TT,
+            (val, is_aux) => {
+                this.#setter(val, is_aux);
```

Comment:
These 7 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2425-ef751130-20251217-170038-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)